### ISSUE TITLE
Wishlist: buffer-local b:gutentags_ctags_executable

### ISSUE THREAD

**Issue by mgedmin** [ludovicchabant/vim-gutentags#139]
## Background

I'm working on a project that has a Python backend and a JavaScript (ES2016) frontend.  I can generate a single tags file for my entire project using exuberant ctags for the backend parts, combined with es-ctags for the frontend parts:

    ctags -R server/
    es-ctags -a -R client/app/ --extension=.js
    es-ctags -a -R client/app/ --extension=.jsx

The only problem is that Gutentags destroys JavaScript tags because it runs exuberant ctags for every file I edit.

I would like to be able to tell Gutentags which ctags executable it should use for every file.  This is similar to #95, except I've both .js and .jsx files with &filetype set to javascript, but I must use different command lines to have es-ctags handle them (`es-ctags filename.jsx` will produce no output if I don't specify `--extension=.jsx`).

## Proposal

Introduce a buffer-local variable `b:gutentags_ctags_executable` that could override the global `g:gutentags_ctags_executable` (or `g:gutentags_ctags_executable_{type}`).  Then I can define a
`gutentags_init_user_func` that would manually set it to `'ctags'` or `'es-ctags --extension=.' . expand('%:e')`.

**Comment by mgedmin** [comment]
Actually, I think I might need both `b:gutentags_ctags_executable` and `b:gutentags_ctags_extra_args`, since I'm not sure it's fine to include arguments in the _executable option.

**Comment by mgedmin** [comment]
Another update: turns out es-ctags accepts regular expressions for the extension, so I could get by with using g:gutentags_tags_executable_javascript = 'es-ctags --extension=jsx?', if only that option allowed me to specify arguments, and not just the binary itself.

**Comment by ludovicchabant** [comment]
Hi!

Yeah it looks like we'll need something to support mixed projects like this eventually but I haven't been able to come up with a design I like that wouldn't make the plugin code unnecessarily complicated.
(and since I have no use for that myself, it's pretty low on my priorities obviously, so there's a good chance this would have to come from a pull request)

For instance, I'm still undecided between solving this at the "executable" level, _i.e._ not changing much in Gutentags, and instead offer some scripts that can pass as Ctags but are effectively running several other Ctags.... or if Gutentags would need to be aware of all that stuff.

I think the problem with your proposal is that if you were to regenerate the whole tags from scratch (`:GutentagsUpdate!`, or what Gutentags does when you open the first file in your project), it would then only run whatever is set in the buffer variables and you would end up with an incomplete tags file, no?

**Comment by ludovicchabant** [comment]
Actually I think it *may* work if the buffer variable (`b:gutentags_ctags_executable`) is only used for incremental updates (_i.e._ updates when the file is saved, not when the file is opened or any other situation), and the overall/global executable is set to a script that runs all 3 commands (`ctags` and the 2 `es-ctags`) ? Is that what you had in mind?

**Comment by mgedmin** [comment]
I'm only interested in incremental updates.  I don't rely on gutentags to build a global tags file for me -- I want that to be an explicit action (`:!ctags -R .`, or a shell script I have that collects all Python virtualenv/buildout package directories and passes them to ctags, or `:!make tags` that I add to my projects), for reasons that are probably off-topic for this issue.

I don't think it's important whether the tag generation happens for a file because I made changes and saved it, or because I opened it (and the timestamp indicates the file is newer than the tags file, or something -- I'm not sure how gutentags decides when to update the tags for a file).  As long as gutentags does the update-tags-for-a-single-file-routine, it's useful to be able to tell it which tags generator to use.

I saw that gutentags had a buffer-local `b:gutentags_ctags_tagfile`, which I think is meant to solve a similar problem (merging tags produced by different generators), and I thought that a similar option might make the solution easier (no need to update my `make tags` to split the tags file into multiple ones; no need to configure vim to load tags from multiple files).

At the moment I went with a ctags wrapper script that dispatches to ctags or es-ctags depending on the extension of the last argument, which means the issue is not pressing to me.  Maybe buffer-local variables are a clunky hack and it would be better to allow filetype-specific ctags executables with parameters (eg. `let g:gutentags_ctags_executable_javascript = ['es-ctags', '--extensions=(js|jsx)']).
